[Model] Add BGNN example

nd7141 opened the pull request:
## Description
A proposal of a new GNN model coming from ICLR 2021: https://openreview.net/forum?id=ebS5NUfoMKL
The model is implemented with DGL: https://github.com/nd7141/bgnn

Reviewers: @jermainewang @BarclayII 

Discussion thread: https://github.com/dmlc/dgl/issues/2603

This is initial proposal of the model. There are two files: `BGNN.py` which implements BGNN model and `run.py` which shows how to use it. 

I also include datasets I used in the paper for the sake of reproducibility (although I think that datasets should be included in DGL as a separate request). 

Current results match those of the papers (although I have not done thorough hyperparameter tuning). 

Here are images showing performance of BGNN, that is similar to the paper (https://openreview.net/forum?id=ebS5NUfoMKL): 
![image](https://user-images.githubusercontent.com/4607709/110802539-914a7c00-827e-11eb-9ab5-69d60c07d449.png)
![image](https://user-images.githubusercontent.com/4607709/110802556-94de0300-827e-11eb-9b16-4832c2f8a799.png)
![image](https://user-images.githubusercontent.com/4607709/110802611-a4f5e280-827e-11eb-8a26-042326575cb8.png)
![image](https://user-images.githubusercontent.com/4607709/110802630-a9220000-827e-11eb-9f0c-44a3df05e4ee.png)
![image](https://user-images.githubusercontent.com/4607709/110802672-b0490e00-827e-11eb-861a-65742d3b6933.png)
![image](https://user-images.githubusercontent.com/4607709/110802737-bdfe9380-827e-11eb-8083-420e73409cdb.png)

I think `BGNN.py` should be part of the DGL, in the sense that an end user can just do `import dgl.BGNNPredictor`. On the other hand, `run.py` can be kept in the example. 

A few notes: 
* Currently `BGNN.py` includes as well plotting, which may not be needed for the final version.
* I evaluate for train, valid, and test masks during training, which can slow down training. These parts could be excluded, but should be discussed as it impacts early stopping. 
* BGNN relies on CatBoost package (https://catboost.ai/), so it should be installed if an end user uses BGNN. 
* There are some packages which could be installed as well but are not necessary (e.g. tqdm, sklearn). 

I am happy to answer your questions and discuss how to integrate BGNN in DGL in the optimal way. Please let me know your thoughts.

BarclayII commented:
Thanks a lot!  It's a brand new model contribution so I put an item in the model index README.

Could you also add a README with instructions on how to download the dataset (once you have uploaded them)?

nd7141 commented:
@BarclayII @jermainewang I updated the code in the example and provided instructions how you can download datasets. Please let me know if something else is needed from me?

jermainewang commented on examples/pytorch/bgnn/BGNN.py:
```diff
@@ -0,0 +1,356 @@
+                 append_gbdt_pred = True,
+                 train_input_features = False,
+                 gbdt_depth=6,
+                 gbdt_lr=0.1):
```
Could you pls add docstring for these? We follow the [numpy doc style](https://numpydoc.readthedocs.io/en/latest/format.html). [Code example in DGL](https://github.com/dmlc/dgl/blob/master/python/dgl/nn/pytorch/conv/graphconv.py#L141). It's good to have some code examples for usage.

nd7141 replied:
Updated.

jermainewang commented on examples/pytorch/bgnn/BGNN.py:
```diff
@@ -0,0 +1,356 @@
+                                  depth=self.gbdt_depth,
+                                  learning_rate=self.gbdt_lr,
+                                  loss_function=catboost_loss_fn,
+                                  random_seed=0,
```
expose `random_seed` to the init function of `BGNNPredictor`?

jermainewang commented on examples/pytorch/bgnn/BGNN.py:
```diff
@@ -0,0 +1,356 @@
+
+        gbdt_X_train = X.iloc[train_mask]
+        gbdt_y_train = y.iloc[train_mask]
+        gbdt_alpha = 1
```
Is this a tunable hyperparameter? Consider expose it to the `__init__` function with default value.

jermainewang commented on examples/pytorch/bgnn/BGNN.py:
```diff
@@ -0,0 +1,356 @@
+            normalize_features=False,
+            replace_na=True,
+            ):
+        '''
```
Same. Pls follow the numpy docstring style.  DGL also has some conventions for docing the argument types:
* Use `dgl.DGLGraph` for DGLGraph.
* Use `numpy.array` for numpy arrays and describe the shape in the description.
* Use `list[int]` for list of integers
* Use python type names for python objects (e.g., `int`, `float`, ...)

jermainewang commented on examples/pytorch/bgnn/BGNN.py:
```diff
@@ -0,0 +1,356 @@
+        '''
+
+        :param graph: dgl graph
+        :param X: numpy matrix of input features. Can be composed of categorical features with missing values.
```
describe how missing values are represented in the matrix

jermainewang commented on examples/pytorch/bgnn/BGNN.py:
```diff
@@ -0,0 +1,356 @@
+        :param graph: dgl graph
+        :param X: numpy matrix of input features. Can be composed of categorical features with missing values.
+        :param y: numpy matrix of target labels. Can be multidimensional targets.
+        :param train_mask: list of integers (rows in X that correspond to train set)
```
describe the element of the mask array (e.g., 0-1 array?)

jermainewang commented on examples/pytorch/bgnn/BGNN.py:
```diff
@@ -0,0 +1,356 @@
+        :param num_epochs: number of epochs to run
+        :param patience: number of epochs to wait until early stopping
+        :param logging_epochs: log every n epoch
+        :param metric_name: metric to use for early stopping
```
I wonder how different early stopping metric performs in practice? Is it critical? Can we remove the option and use the default value `'loss'`?

nd7141 replied:
I think we better keep as it is, as we want to look at accuracy and not at the loss, if accuracy if our final goal.

jermainewang commented on examples/pytorch/bgnn/BGNN.py:
```diff
@@ -0,0 +1,356 @@
+        :param logging_epochs: log every n epoch
+        :param metric_name: metric to use for early stopping
+        :param normalize_features: normalize original input features X (column wise)
+        :param replace_na: replace NA in X
```
I wonder if these options (`normalize_features` and `replace_na`) should be here. They look like data preprocessing steps that may go to the example code.

nd7141 replied:
Moved them to run.py for preprocessing.

nd7141 commented:
Hey @jermainewang 

I updated the code documentation and also moved the preprocessing to run.py (that requires adding one additional parameter for fit function, as GBDT still want to operate on non-preprocessed features). That worked on my side. I hope it works for you too.

jermainewang reviewed (approved):
LGTM. Thanks for the contribution!

jermainewang commented:
@BarclayII do you have other comments?

BarclayII commented:
I'm good.  Thanks!
